File: WinXound_Net/wxImportExport.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.IO;
using System.Windows.Forms;

using ScintillaTextEditor;


namespace WinXound_Net
{
    class wxImportExport
    {

        private string newline = System.Environment.NewLine;

        private string mWorkingDirectory = null;
        public void SetWorkingDirectory(string WorkingDirectory)
        {
            mWorkingDirectory = WorkingDirectory;
        }

        public wxImportExport(string WorkingDirectory)
        {
            mWorkingDirectory = WorkingDirectory;
        }



        ////////////////////////////////////////////////////////////////////////////////////
        //IMPORT CLASSES

        //ORC/SCO
        public string ImportORCSCO(ref string SciEditFileName)
        {
            try
            {
                string tempOrc = "";
                string tempSco = "";

                string OrcFileName = SciEditFileName.Remove(SciEditFileName.Length - 4, 4) + ".orc";
                string ScoFileName = SciEditFileName.Remove(SciEditFileName.Length - 4, 4) + ".sco";

                //Verify that files exists and load them 
                if (File.Exists(OrcFileName))
                {
                    tempOrc = File.ReadAllText(OrcFileName, System.Text.Encoding.Default);
                }
                if (File.Exists(ScoFileName))
                {
                    tempSco = File.ReadAllText(ScoFileName, System.Text.Encoding.Default);
                }

                string tempString = "<CsoundSynthesizer>" + newline + newline +
                                    "<CsOptions>" + newline + 
                                    "</CsOptions>" + newline + newline +
                                    "<CsInstruments>" + newline + newline +
                                    tempOrc + newline + newline +
                                    "</CsInstruments>" + newline + newline +
                                    "<CsScore>" + newline + newline +
                                    tempSco + newline + newline +
                                    "</CsScore>" + newline + newline +
                                    "</CsoundSynthesizer>" + newline + newline;

                ////SciEditFileName = SciEditFileName.Remove(SciEditFileName.Length - 4, 4) + ".csd";
                SciEditFileName = Path.GetFileNameWithoutExtension(SciEditFileName) + ".csd";

                return tempString;
            }
            catch (Exception ex)
            {
                wxGlobal.wxMessageError(ex.Message, "wxImportExport - ImportORCSCO");
                return "";
            }

        }

        //ORC
        public void ImportORC(TextEditor textEditor, string OrcFileName)
        {
            try
            {
                string tempOrc = File.ReadAllText(OrcFileName, System.Text.Encoding.Default);
                if (string.IsNullOrEmpty(tempOrc)) return;

                tempOrc = tempOrc.Insert(0, newline + newline);
                tempOrc += newline;

                Int32 startInstr = 0;
                Int32 endInstr = 0;

                //Search for <CsInstruments> section: start and end
                //Find <CsInstruments> 
                startInstr = textEditor.FindText("<CsInstruments>", true, true, false, false, false, true, 0, -1);
                if (startInstr == -1)
                {
                    MessageBox.Show("<CsInstruments> tag not found!" + newline +
                                    "Please insert it in the code and retry.", 
                                    "Import Orc Error!",
                                    MessageBoxButtons.OK,
                                    MessageBoxIcon.Error);
                    return;
                    //tempOrc.Insert(0, "<CsInstruments>");
                    //startInstr = textEditor.GetCaretPosition();
                }

                //Find </CsInstruments> 
                endInstr = textEditor.FindText("</CsInstruments>", true, true, false, false, false, true, 0, -1);
                if (endInstr == -1)
                {
                    MessageBox.Show("</CsInstruments> tag not found!" + newline +
                                    "Please insert it in the code and retry.",
                                    "Import Orc Error!",
                                    MessageBoxButtons.OK,
                                    MessageBoxIcon.Error);
                    return;
                    //tempOrc += "</CsInstruments>";
                    //endInstr = startInstr;
                }

                textEditor.SetSelection(startInstr + "<CsInstruments>".Length, endInstr);

                //Replace/Insert text
                //textEditor.PrimaryView.BeginUndoAction();
                textEditor.SetSelectedText(textEditor.ConvertEolOfString(tempOrc));
                //textEditor.ConvertEOL(textEditor.GetEolMode());
                //textEditor.PrimaryView.EndUndoAction();

            }
            catch (Exception ex)
            {
                wxGlobal.wxMessageError(ex.Message, "wxImportExport - ImportORC");
            }

        }


        //SCO
        public void ImportSCO(TextEditor textEditor, string ScoFileName)
        {
            try
            {
                string tempSco = File.ReadAllText(ScoFileName, System.Text.Encoding.Default);
                if (string.IsNullOrEmpty(tempSco)) return;

                tempSco = tempSco.Insert(0, newline + newline);
                tempSco += newline;

                Int32 startScore = 0;
                Int32 endScore = 0;

                //Search for <CsScore> section: start and end
                //Find <CsScore> 
                startScore = textEditor.FindText("<CsScore>", true, true, false, false, false, true, 0, -1);
                if (startScore == -1)
                {
                    MessageBox.Show("<CsScore> tag not found!" + newline +
                                    "Please insert it in the code and retry.",
                                    "Import Sco Error!",
                                    MessageBoxButtons.OK,
                                    MessageBoxIcon.Error);
                    return;
                }

                //Find </CsScore> 
                endScore = textEditor.FindText("</CsScore>", true, true, false, false, false, true, 0, -1);
                if (endScore == -1)
                {
                    MessageBox.Show("</CsScore> tag not found!" + newline +
                                    "Please insert it in the code and retry.",
                                    "Import Sco Error!",
                                    MessageBoxButtons.OK,
                                    MessageBoxIcon.Error);
                    return;
                }

                textEditor.SetSelection(startScore + "<CsScore>".Length, endScore);

                //Replace/Insert text
                //textEditor.PrimaryView.BeginUndoAction();
                textEditor.SetSelectedText(textEditor.ConvertEolOfString(tempSco));
                //textEditor.ConvertEOL(textEditor.GetEolMode());
                //textEditor.PrimaryView.EndUndoAction();

            }
            catch (Exception ex)
            {
                wxGlobal.wxMessageError(ex.Message, "wxImportExport - ImportSCO");
            }
        }




        ////////////////////////////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////////////////////////////
        ////////////////////////////////////////////////////////////////////////////////////




        ////////////////////////////////////////////////////////////////////////////////////
        // EXPORT CLASSES
        public void ExportOrcSco(string mText, string SciEditFileName)
        {

            int InizioOrchestra = 0;
            int FineOrchestra = 0;
            int InizioScore = 0;
            int FineScore = 0;

            RichTextBox RichTextBox1 = new RichTextBox();
            RichTextBox RichTextBox2 = new RichTextBox();
            RichTextBox RichTextBox3 = new RichTextBox();

            RichTextBox1.Text = mText;
            RichTextBox2.Text = "";
            RichTextBox3.Text = "";

            //Find <CsInstruments> tags 
            try
            {
                InizioOrchestra = RichTextBox1.Find("<CsInstruments>", 0, RichTextBox1.Text.Length,
                                  RichTextBoxFinds.WholeWord | RichTextBoxFinds.MatchCase);
                if (!(InizioOrchestra > 0))
                    throw new Exception("Unable to find <CsInstruments> section");

                FineOrchestra = RichTextBox1.Find("</CsInstruments>", 0, RichTextBox1.Text.Length,
                                RichTextBoxFinds.WholeWord | RichTextBoxFinds.MatchCase);
                if (!(FineOrchestra > 0))
                    throw new Exception("Unable to find </CsInstruments> section");
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Error!",
                                MessageBoxButtons.OK,
                                MessageBoxIcon.Error);
                RichTextBox1.Dispose();
                RichTextBox2.Dispose();
                RichTextBox3.Dispose();
                return;
            }

            RichTextBox1.SelectionStart = InizioOrchestra + "<CsInstruments>".Length;
            RichTextBox1.SelectionLength = FineOrchestra - RichTextBox1.SelectionStart;

            //Set Text in RTB2 
            RichTextBox2.AppendText(RichTextBox1.SelectedText.Trim());

            //Deselect text in richtextbox 
            RichTextBox1.SelectionLength = 0;
            RichTextBox1.SelectionStart = 0;


            //Find <CsScore> tags 
            try
            {
                InizioScore = RichTextBox1.Find("<CsScore>", 0, RichTextBox1.Text.Length,
                                                RichTextBoxFinds.WholeWord | RichTextBoxFinds.MatchCase);
                if (!(InizioScore > 0))
                    throw new Exception("Unable to find <CsScore> section");

                FineScore = RichTextBox1.Find("</CsScore>", 0, RichTextBox1.Text.Length,
                                                RichTextBoxFinds.WholeWord | RichTextBoxFinds.MatchCase);
                if (!(FineScore > 0))
                    throw new Exception("Unable to find </CsScore> section");

                RichTextBox1.SelectionStart = InizioScore + "<CsScore>".Length;
                RichTextBox1.SelectionLength = FineScore - RichTextBox1.SelectionStart;
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Error!",
                                MessageBoxButtons.OK,
                                MessageBoxIcon.Error);
                RichTextBox1.Dispose();
                RichTextBox2.Dispose();
                RichTextBox3.Dispose();
                return;
            }

            //Set Text in RTB3 
            RichTextBox3.AppendText(RichTextBox1.SelectedText.Trim());



            //Save Files as ORC and SCO 
            SaveFileDialog saveFileDialog1 = new SaveFileDialog();
            saveFileDialog1.FileName = Path.GetFileNameWithoutExtension(SciEditFileName);
            saveFileDialog1.RestoreDirectory = true;
 
            if ((saveFileDialog1.ShowDialog() == DialogResult.OK) &&
                (saveFileDialog1.FileName.Length) > 0)
            {

                RichTextBox2.SaveFile(saveFileDialog1.FileName + ".orc", RichTextBoxStreamType.PlainText);
                RichTextBox3.SaveFile(saveFileDialog1.FileName + ".sco", RichTextBoxStreamType.PlainText);

                MessageBox.Show("Export done!", "Export file ...",
                                MessageBoxButtons.OK,
                                MessageBoxIcon.Information);
            }

            RichTextBox1.Dispose();
            RichTextBox2.Dispose();
            RichTextBox3.Dispose();

        }


        public void ExportORC(string mText, string OrcFileName)
        {

            RichTextBox tempOrc = new RichTextBox();
            RichTextBox tempEditor = new RichTextBox();
            tempEditor.Text = mText;

            string StringToFind = null;
            Int32 mFindPos = -1;
            Int32 mFindPosStart = 0;
            Int32 mFindPosEnd = 0;

            //Find <CsInstruments> 
            StringToFind = "<CsInstruments>";
            mFindPos = tempEditor.Find(StringToFind, 0, tempEditor.Text.Length,
                                       RichTextBoxFinds.WholeWord | RichTextBoxFinds.MatchCase);
            if (mFindPos > -1)
            {
                mFindPosStart = mFindPos + StringToFind.Length;
            }
            else
            {
                MessageBox.Show("Unable to find <CsInstruments> section", "Error",
                                MessageBoxButtons.OK);
                return;
            }

            //Find </CsInstruments> 
            mFindPos = -1;
            StringToFind = "</CsInstruments>";
            mFindPos = tempEditor.Find(StringToFind, 0, tempEditor.Text.Length,
                                       RichTextBoxFinds.WholeWord | RichTextBoxFinds.MatchCase);
            if (mFindPos > -1)
            {
                mFindPosEnd = mFindPos;
            }
            else
            {
                MessageBox.Show("Unable to find </CsInstruments> section", "Error",
                                MessageBoxButtons.OK);
                return;
            }


            tempOrc.AppendText(tempEditor.Text.Substring(mFindPosStart, mFindPosEnd - mFindPosStart).Trim());


            //Save Files as ORC
            SaveFileDialog saveFileDialog1 = new SaveFileDialog();
            saveFileDialog1.FileName = Path.GetFileNameWithoutExtension(OrcFileName);
            saveFileDialog1.Filter = "CSound Orc Files (*.orc)|*.orc|All files (*.*)|*.*";
            saveFileDialog1.FilterIndex = 1;
            saveFileDialog1.RestoreDirectory = true;

            if ((saveFileDialog1.ShowDialog() == DialogResult.OK) &&
                (saveFileDialog1.FileName.Length) > 0)
            {

                tempOrc.SaveFile(saveFileDialog1.FileName, RichTextBoxStreamType.PlainText);
                MessageBox.Show("Export done!", "Export file ...",
                                MessageBoxButtons.OK,
                                MessageBoxIcon.Information);
            }

            tempOrc.Dispose();
            tempEditor.Dispose();

        }


        public void ExportSCO(string mText, string ScoFileName)
        {

            RichTextBox tempSco = new RichTextBox();
            RichTextBox tempEditor = new RichTextBox();
            tempEditor.Text = mText;

            string StringToFind = null;
            Int32 mFindPos = -1;
            Int32 mFindPosStart = 0;
            Int32 mFindPosEnd = 0;

            //Find <CsInstruments> 
            StringToFind = "<CsScore>";
            mFindPos = tempEditor.Find(StringToFind, 0, tempEditor.Text.Length,
                                       RichTextBoxFinds.WholeWord | RichTextBoxFinds.MatchCase);
            if (mFindPos > -1)
            {
                mFindPosStart = mFindPos + StringToFind.Length;
            }
            else
            {
                MessageBox.Show("Unable to find <CsScore> section", "Error",
                                MessageBoxButtons.OK);
                return;
            }

            //Find </CsInstruments> 
            mFindPos = -1;
            StringToFind = "</CsScore>";
            mFindPos = tempEditor.Find(StringToFind, 0, tempEditor.Text.Length,
                                       RichTextBoxFinds.WholeWord | RichTextBoxFinds.MatchCase);
            if (mFindPos > -1)
            {
                mFindPosEnd = mFindPos;
            }
            else
            {
                MessageBox.Show("Unable to find </CsScore> section", "Error",
                                MessageBoxButtons.OK);
                return;
            }


            tempSco.AppendText(tempEditor.Text.Substring(mFindPosStart, mFindPosEnd - mFindPosStart).Trim());


            //Save Files as ORC
            SaveFileDialog saveFileDialog1 = new SaveFileDialog();
            saveFileDialog1.FileName = Path.GetFileNameWithoutExtension(ScoFileName);
            saveFileDialog1.Filter = "CSound Sco Files (*.sco)|*.sco|All files (*.*)|*.*";
            saveFileDialog1.FilterIndex = 1;
            saveFileDialog1.RestoreDirectory = true;
 
            if ((saveFileDialog1.ShowDialog() == DialogResult.OK) &&
                (saveFileDialog1.FileName.Length) > 0)
            {

                tempSco.SaveFile(saveFileDialog1.FileName, RichTextBoxStreamType.PlainText);
                MessageBox.Show("Export done!", "Export file ...",
                                MessageBoxButtons.OK,
                                MessageBoxIcon.Information);
            }

            tempSco.Dispose();
            tempEditor.Dispose();

        }



    }//End (Class)

}

File: src/wx-importexport.cc
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*- */
/*
 * winxound_gtkmm
 * Copyright (C) Stefano Bonetti 2010 <stefano_bonetti@tin.it>
 * 
 * winxound_gtkmm is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * winxound_gtkmm is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "wx-importexport.h"
#include "wx-global.h"
#include "wx-settings.h"
#include "wx-textEditor.h"


wxImportExport::wxImportExport() 	
{

}
wxImportExport::~wxImportExport(void)
{	

}







Glib::ustring wxImportExport::ImportORCSCO(const gchar* filename)
{
	try
	{
		Glib::ustring tempOrc = "";
		Glib::ustring tempSco = "";
		Glib::ustring SciEditFileName = filename;

		//string OrcFileName = SciEditFileName.Remove(SciEditFileName.Length - 4, 4) + ".orc";
		//string ScoFileName = SciEditFileName.Remove(SciEditFileName.Length - 4, 4) + ".sco";
		Glib::ustring OrcFileName = SciEditFileName.substr(0, SciEditFileName.size() - 4);
		OrcFileName.append(".orc");
		Glib::ustring ScoFileName = SciEditFileName.substr(0, SciEditFileName.size() - 4);
		ScoFileName.append(".sco");

		
		//Verify that files exists and load them 
		if (Glib::file_test(OrcFileName, Glib::FILE_TEST_EXISTS))
		{
			//tempOrc = File.ReadAllText(OrcFileName, System.Text.Encoding.Default);
			tempOrc = Glib::file_get_contents(OrcFileName);
		}
		if (Glib::file_test(ScoFileName, Glib::FILE_TEST_EXISTS))
		{
			//tempSco = File.ReadAllText(ScoFileName, System.Text.Encoding.Default);
			tempSco = Glib::file_get_contents(ScoFileName);
		}

		Glib::ustring tempString = 
			Glib::ustring::compose("<CsoundSynthesizer>\n\n"
			                       "<CsOptions>\n"
			                       "</CsOptions>\n\n"
			                       "<CsInstruments>\n\n"
			                       "%1\n\n" //tempOrc + newline + newline +
			                       "</CsInstruments>\n\n"
			                       "<CsScore>\n\n"
			                       "%2\n\n" //tempSco + newline + newline +
			                       "</CsScore>\n\n"
			                       "</CsoundSynthesizer>\n\n", tempOrc, tempSco);

		
		////SciEditFileName = SciEditFileName.Remove(SciEditFileName.Length - 4, 4) + ".csd";
		//SciEditFileName = Path.GetFileNameWithoutExtension(SciEditFileName) + ".csd";
		//Glib::ustring ret = SciEditFileName.substr(0, SciEditFileName.size() - 4);
		//ret.append(".csd");

			
		return tempString;
		
	}
	catch (...)
	{
		wxGLOBAL->DebugPrint("wxImportExport - ImportORCSCO Error ");
	}

	return "";

}



//ORC
void wxImportExport::ImportORC(wxTextEditor* textEditor, const gchar* OrcFileName)
{
	try
	{
		Glib::ustring tempOrc = Glib::file_get_contents(OrcFileName);
		if (tempOrc == "") return;

		//tempOrc = tempOrc.Insert(0, newline + newline);
		tempOrc.insert(0, "\n\n");
		tempOrc.append("\n");

		//wxGLOBAL->DebugPrint("", tempOrc.c_str());
		
		gint startInstr = 0;
		gint endInstr = 0;

		//Search for <CsInstruments> section: start and end
		//Find <CsInstruments> 
		startInstr = 
			textEditor->FindText("<CsInstruments>", true, true, false, false, false, true, 0, -1);
		if (startInstr == -1)
		{
			wxGLOBAL->ShowMessageBox("<CsInstruments> tag not found!\n"
			                         "Please insert it in the code and retry.", 
			                         "Import Orc Error!",
			                         Gtk::BUTTONS_OK);
			return;
		}

		//Find </CsInstruments> 
		endInstr = 
			textEditor->FindText("</CsInstruments>", true, true, false, false, false, true, 0, -1);
		if (endInstr == -1)
		{
			wxGLOBAL->ShowMessageBox("</CsInstruments> tag not found!\n"
			                         "Please insert it in the code and retry.",
			                         "Import Orc Error!",
			                         Gtk::BUTTONS_OK);
			return;
		}

		textEditor->setSelection(startInstr + strlen("<CsInstruments>"), endInstr);

		//Replace/Insert text
		Glib::ustring temp = textEditor->ConvertEolOfString(tempOrc.c_str());
		textEditor->setSelectedText(temp.c_str());
	}
	catch (...)
	{
		wxGLOBAL->DebugPrint("wxImportExport - ImportORC Error ");
	}

}


//SCO
void wxImportExport::ImportSCO(wxTextEditor* textEditor, const gchar* ScoFileName)
{


	try
	{
		Glib::ustring tempSco = Glib::file_get_contents(ScoFileName);
		if (tempSco == "") return;

		//tempSco = tempSco.Insert(0, newline + newline);
		tempSco.insert(0, "\n\n");
		tempSco.append("\n");

		//wxGLOBAL->DebugPrint("", tempSco.c_str());
		
		gint startScore = 0;
		gint endScore = 0;

		//Search for <CsScore> section: start and end
		//Find <CsScore> 
		startScore = 
			textEditor->FindText("<CsScore>", true, true, false, false, false, true, 0, -1);
		if (startScore == -1)
		{
			wxGLOBAL->ShowMessageBox("<CsScore> tag not found!\n"
			                         "Please insert it in the code and retry.", 
			                         "Import Sco Error!",
			                         Gtk::BUTTONS_OK);
			return;
		}

		//Find </CsScore> 
		endScore =
			textEditor->FindText("</CsScore>", true, true, false, false, false, true, 0, -1);
		if (endScore == -1)
		{
			wxGLOBAL->ShowMessageBox("</CsScore> tag not found!\n"
			                         "Please insert it in the code and retry.",
			                         "Import Sco Error!",
			                         Gtk::BUTTONS_OK);
			return;
		}

		textEditor->setSelection(startScore + strlen("<CsScore>"), endScore);

		//Replace/Insert text
		Glib::ustring temp = textEditor->ConvertEolOfString(tempSco.c_str());
		textEditor->setSelectedText(temp.c_str());
	}
	catch (...)
	{
		wxGLOBAL->DebugPrint("wxImportExport - ImportSCO Error ");
	}

}




////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////




////////////////////////////////////////////////////////////////////////////////////
// EXPORT CLASSES
void wxImportExport::ExportOrcSco(wxTextEditor* textEditor, const gchar* filename)
{

	gint startInstr = 0;
	gint endInstr = 0;

	Glib::ustring tempOrc = "";
	Glib::ustring tempSco = "";

	//Search for <CsInstruments> section: start and end
	//Find <CsInstruments> 
	startInstr = 
		textEditor->FindText("<CsInstruments>", true, true, false, false, false, true, 0, -1);
	if (startInstr == -1)
	{
		wxGLOBAL->ShowMessageBox("<CsInstruments> tag not found!\n"
		                         "Please insert it in the code and retry.", 
		                         "Export Orc/Sco Error!",
		                         Gtk::BUTTONS_OK);
		return;
	}

	//Find </CsInstruments> 
	endInstr = 
		textEditor->FindText("</CsInstruments>", true, true, false, false, false, true, 0, -1);
	if (endInstr == -1)
	{
		wxGLOBAL->ShowMessageBox("</CsInstruments> tag not found!\n"
		                         "Please insert it in the code and retry.",
		                         "Export Orc/Sco Error!",
		                         Gtk::BUTTONS_OK);
		return;
	}



	
	gint startScore = 0;
	gint endScore = 0;

	//Search for <CsScore> section: start and end
	//Find <CsScore> 
	startScore = 
		textEditor->FindText("<CsScore>", true, true, false, false, false, true, 0, -1);
	if (startScore == -1)
	{
		wxGLOBAL->ShowMessageBox("<CsScore> tag not found!\n"
		                         "Please insert it in the code and retry.", 
		                         "Export Orc/Sco Error!",
		                         Gtk::BUTTONS_OK);
		return;
	}

	//Find </CsScore> 
	endScore =
		textEditor->FindText("</CsScore>", true, true, false, false, false, true, 0, -1);
	if (endScore == -1)
	{
		wxGLOBAL->ShowMessageBox("</CsScore> tag not found!\n"
		                         "Please insert it in the code and retry.",
		                         "Export Orc/Sco Error!",
		                         Gtk::BUTTONS_OK);
		return;
	}


	//Retrieve text
	tempOrc = textEditor->getTextRange(startInstr + strlen("<CsInstruments>"), endInstr);
	tempSco = textEditor->getTextRange(startScore + strlen("<CsScore>"), endScore);
	


	Gtk::FileChooserDialog dialog("Export CSD File to Orc/Sco",
	                              Gtk::FILE_CHOOSER_ACTION_SAVE);
	//dialog.set_transient_for(*this);
	dialog.set_modal(TRUE);
	dialog.set_current_folder(wxSETTINGS->Directory.LastUsedPath);

	//Add response buttons the the dialog:
	dialog.add_button(Gtk::Stock::CANCEL, Gtk::RESPONSE_CANCEL);
	dialog.add_button(Gtk::Stock::SAVE, Gtk::RESPONSE_OK);

	/*
	Gtk::FileFilter filter_supported_files;
	filter_supported_files.set_name("CSD files");
	//filter_supported_files.add_mime_type("text/plain");
	filter_supported_files.add_pattern("*.csd");
	dialog.add_filter(filter_supported_files);
	*/
	 

	Glib::ustring t1 = Glib::path_get_basename(filename);
	Glib::ustring t2 = t1.substr(0, t1.size() - 4);
	dialog.set_current_name(t2);

	//If the WorkingDir is not empty and exists add it to the Open Dialog Box:
	if(Glib::file_test(wxSETTINGS->Directory.WorkingDir, 
	                   (Glib::FILE_TEST_EXISTS | Glib::FILE_TEST_IS_DIR)))
	{
		dialog.add_shortcut_folder(wxSETTINGS->Directory.WorkingDir);
	}
	

	int result = dialog.run();

	if (result != Gtk::RESPONSE_OK)
	{
		return;
	}

	//ActiveEditor.SaveFile(ActiveEditor.FileName);
	std::string file = dialog.get_filename();
	wxSETTINGS->Directory.LastUsedPath = dialog.get_current_folder();

	
	Glib::file_set_contents(Glib::ustring::compose("%1.orc",file), wxGLOBAL->Trim(tempOrc));
	Glib::file_set_contents(Glib::ustring::compose("%1.sco",file), wxGLOBAL->Trim(tempSco));

	wxGLOBAL->ShowMessageBox("Export done!",
	                         "Export Orc/Sco",
	                         Gtk::BUTTONS_OK);

}



void wxImportExport::ExportORC(wxTextEditor* textEditor, const gchar* filename)
{

	gint startInstr = 0;
	gint endInstr = 0;

	Glib::ustring tempOrc = "";

	//Search for <CsInstruments> section: start and end
	//Find <CsInstruments> 
	startInstr = 
		textEditor->FindText("<CsInstruments>", true, true, false, false, false, true, 0, -1);
	if (startInstr == -1)
	{
		wxGLOBAL->ShowMessageBox("<CsInstruments> tag not found!\n"
		                         "Please insert it in the code and retry.", 
		                         "Export Orc/Sco Error!",
		                         Gtk::BUTTONS_OK);
		return;
	}

	//Find </CsInstruments> 
	endInstr = 
		textEditor->FindText("</CsInstruments>", true, true, false, false, false, true, 0, -1);
	if (endInstr == -1)
	{
		wxGLOBAL->ShowMessageBox("</CsInstruments> tag not found!\n"
		                         "Please insert it in the code and retry.",
		                         "Export Orc/Sco Error!",
		                         Gtk::BUTTONS_OK);
		return;
	}


	//Retrieve text
	tempOrc = textEditor->getTextRange(startInstr + strlen("<CsInstruments>"), endInstr);
	
	Gtk::FileChooserDialog dialog("Export Orchestra Section",
	                              Gtk::FILE_CHOOSER_ACTION_SAVE);
	//dialog.set_transient_for(*this);
	dialog.set_modal(TRUE);
	dialog.set_current_folder(wxSETTINGS->Directory.LastUsedPath);

	//Add response buttons the the dialog:
	dialog.add_button(Gtk::Stock::CANCEL, Gtk::RESPONSE_CANCEL);
	dialog.add_button(Gtk::Stock::SAVE, Gtk::RESPONSE_OK);

	/*
	Gtk::FileFilter filter_supported_files;
	filter_supported_files.set_name("CSD files");
	//filter_supported_files.add_mime_type("text/plain");
	filter_supported_files.add_pattern("*.csd");
	dialog.add_filter(filter_supported_files);
	*/


	Glib::ustring t1 = Glib::path_get_basename(filename);
	Glib::ustring t2 = t1.substr(0, t1.size() - 4);
	dialog.set_current_name(t2);

	//If the WorkingDir is not empty and exists add it to the Open Dialog Box:
	if(Glib::file_test(wxSETTINGS->Directory.WorkingDir, 
	                   (Glib::FILE_TEST_EXISTS | Glib::FILE_TEST_IS_DIR)))
	{
		dialog.add_shortcut_folder(wxSETTINGS->Directory.WorkingDir);
	}

	
	
	int result = dialog.run();

	if (result != Gtk::RESPONSE_OK)
	{
		return;
	}

	//ActiveEditor.SaveFile(ActiveEditor.FileName);
	std::string file = dialog.get_filename();
	wxSETTINGS->Directory.LastUsedPath = dialog.get_current_folder();

	Glib::file_set_contents(Glib::ustring::compose("%1.orc",file), wxGLOBAL->Trim(tempOrc));


	wxGLOBAL->ShowMessageBox("Export done!",
	                         "Export Orc",
	                         Gtk::BUTTONS_OK);	

}


void wxImportExport::ExportSCO(wxTextEditor* textEditor, const gchar* filename)
{
	gint startInstr = 0;
	gint endInstr = 0;

	Glib::ustring tempSco = "";
	
	gint startScore = 0;
	gint endScore = 0;

	//Search for <CsScore> section: start and end
	//Find <CsScore> 
	startScore = 
		textEditor->FindText("<CsScore>", true, true, false, false, false, true, 0, -1);
	if (startScore == -1)
	{
		wxGLOBAL->ShowMessageBox("<CsScore> tag not found!\n"
		                         "Please insert it in the code and retry.", 
		                         "Export Orc/Sco Error!",
		                         Gtk::BUTTONS_OK);
		return;
	}

	//Find </CsScore> 
	endScore =
		textEditor->FindText("</CsScore>", true, true, false, false, false, true, 0, -1);
	if (endScore == -1)
	{
		wxGLOBAL->ShowMessageBox("</CsScore> tag not found!\n"
		                         "Please insert it in the code and retry.",
		                         "Export Orc/Sco Error!",
		                         Gtk::BUTTONS_OK);
		return;
	}


	//Retrieve text
	tempSco = textEditor->getTextRange(startScore + strlen("<CsScore>"), endScore);
	
	Gtk::FileChooserDialog dialog("Export Score Section",
	                              Gtk::FILE_CHOOSER_ACTION_SAVE);
	//dialog.set_transient_for(*this);
	dialog.set_modal(TRUE);
	dialog.set_current_folder(wxSETTINGS->Directory.LastUsedPath);

	//Add response buttons the the dialog:
	dialog.add_button(Gtk::Stock::CANCEL, Gtk::RESPONSE_CANCEL);
	dialog.add_button(Gtk::Stock::SAVE, Gtk::RESPONSE_OK);

	/*
	Gtk::FileFilter filter_supported_files;
	filter_supported_files.set_name("CSD files");
	//filter_supported_files.add_mime_type("text/plain");
	filter_supported_files.add_pattern("*.csd");
	dialog.add_filter(filter_supported_files);
	*/
	 

	Glib::ustring t1 = Glib::path_get_basename(filename);
	Glib::ustring t2 = t1.substr(0, t1.size() - 4);
	dialog.set_current_name(t2);

	//If the WorkingDir is not empty and exists add it to the Open Dialog Box:
	if(Glib::file_test(wxSETTINGS->Directory.WorkingDir, 
	                   (Glib::FILE_TEST_EXISTS | Glib::FILE_TEST_IS_DIR)))
	{
		dialog.add_shortcut_folder(wxSETTINGS->Directory.WorkingDir);
	}

	
	
	int result = dialog.run();

	if (result != Gtk::RESPONSE_OK)
	{
		return;
	}

	//ActiveEditor.SaveFile(ActiveEditor.FileName);
	std::string file = dialog.get_filename();
	wxSETTINGS->Directory.LastUsedPath = dialog.get_current_folder();

	Glib::file_set_contents(Glib::ustring::compose("%1.sco",file), wxGLOBAL->Trim(tempSco));

	wxGLOBAL->ShowMessageBox("Export done!",
	                         "Export Sco",
	                         Gtk::BUTTONS_OK);	

}

